### PR TITLE
zbig01 user interface(key word is responsive) improvements set 1

### PR DESCRIPTION
**Pull request by bradymiller**
First set of user interface script by zbig01 rebased to most recent codebase. zbig01 also has a new stylesheet here:
Administration->Globals->Appearance->General Theme->Flat Red

here's the original PR:
https://github.com/openemr/openemr/pull/938

### CODE REVIEW

**Comment by bradymiller** [comment]
@zbig01 ,
Here is the first installment :)
This should be in a good form for testing/review of your code by the community.
-brady

File: interface/forms/soap/templates/general_new.html [hunk 1]
```diff
@@ -4,49 +4,79 @@
 <title>{xl t='SOAP'|escape:'html'}</title>
 
 {headerTemplate}
+{literal}
+ <style type="text/css" title="mystyles" media="all">
+ <!--
```
Comment (bradymiller) [inline interface/forms/soap/templates/general_new.html]: @zbig01 , Is this entire section commented out?

File: interface/forms/care_plan/new.php [hunk 2]
```diff
@@ -57,16 +58,31 @@
...
-
-        <script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-min-3-1-1/index.js"></script>
-        <script type="text/javascript" src="<?php echo $GLOBALS['webroot'] ?>/library/textformat.js?v=<?php echo $v_js_includes; ?>"></script>
-        <script type="text/javascript" src="<?php echo $GLOBALS['webroot'] ?>/library/dialog.js?v=<?php echo $v_js_includes; ?>"></script>
-        <script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-datetimepicker-2-5-4/build/jquery.datetimepicker.full.min.js"></script>
-
-        <link rel="stylesheet" href="<?php echo $css_header; ?>" type="text/css">
-        <link rel="stylesheet" href="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-datetimepicker-2-5-4/build/jquery.datetimepicker.min.css">
-
+        <?php Header::setupHeader(['bootstrap', 'datetime-picker']);?>
```
Comment (robertdown) [inline interface/forms/care_plan/new.php]: Bootstrap is autoloaded in the Header, no need to call it

File: interface/forms/care_plan/new.php [hunk 3]
```diff
@@ -57,16 +58,31 @@
...
+        <style type="text/css" title="mystyles" media="all">
```
Comment (robertdown) [inline interface/forms/care_plan/new.php]: What's the rationale for adding styles here instead of in the actual theme?

File: interface/forms/care_plan/new.php [hunk 4]
```diff
@@ -148,99 +166,111 @@ function set_related(codetype, code, selector, codedesc) {
...
-                foreach ($check_res as $key => $obj) {
+        <div class="container">
+
+            <div class="row">
+                <div class="page-header">
+                        <h2><?php echo xlt('Care Plan Form'); ?></h2>
+                </div>
+            </div>
+            <div class="row">
+            <?php echo "<form method='post' name='my_form' " . "action='$rootdir/forms/care_plan/save.php?id=" . attr($formid) . "'>\n"; ?>
```
Comment (robertdown) [inline interface/forms/care_plan/new.php]: Would be better to break out the `<form>` to regular HTML and just inject the PHP echo where needed, even it that means multiple time on the line. Or just create a separate line for the action path and do 1 echo

File: interface/forms/care_plan/new.php [hunk 5]
```diff
@@ -148,99 +166,111 @@ function set_related(codetype, code, selector, codedesc) {
...
+                    <?php
+                    if (!empty($check_res)) {
+                        foreach ($check_res as $key => $obj) {
                     ?>
-                    <tr class="tb_row" id="tb_row_<?php echo attr($key) + 1; ?>">
-                        <td align="left" class="forms"><?php echo xlt('Code'); ?>:</td>
-                        <td class="forms">
-                            <input type="text" id="code_<?php echo attr($key) + 1; ?>" style="width:210px" name="code[]" class="code" value="<?php echo text($obj{"code"}); ?>"  onclick='sel_code(this.parentElement.parentElement.id);'><br>
-                            <span id="displaytext_<?php echo attr($key) + 1; ?>" style="width:210px !important;display: block;font-size:13px;color: blue;" class="displaytext"><?php echo text($obj{"codetext"}); ?></span>
+                    <div class = "tb_row" id="tb_row_<?php echo attr($key) + 1; ?>">
```
Comment (robertdown) [inline interface/forms/care_plan/new.php]: This will fail the code standards, should be `class="tb_row"`

File: interface/forms/soap/templates/general_new.html (hunk 1 above)
Comment (zbig01) [inline interface/forms/soap/templates/general_new.html]: No, This internal stylesheet gives the background color for the legend and fieldset for all non flat stylesheets. You will find this internal stylesheet in all the forms that I have modified.  Once a reasonable number of pages have been restyled this stylesheet along with any other modification can be made into an openemr specific override of the default bootstrap styles.
The html style block comment is invalid in css.

File: interface/forms/clinical_instructions/new.php [hunk 6]
```diff
@@ -36,28 +37,61 @@
 ?>
 <html>
     <head>
-        <link rel="stylesheet" href="<?php echo $css_header; ?>" type="text/css">
+        <?php Header::setupHeader('bootstrap'); ?>
```
Comment (robertdown) [inline interface/forms/clinical_instructions/new.php]: bootstrap is autoloaded, no need to include it manually

File: interface/forms/clinical_instructions/new.php [hunk 7]
```diff
@@ -36,28 +37,61 @@
...
+        
+        <style type="text/css" title="mystyles" media="all">
```
Comment (robertdown) [inline interface/forms/clinical_instructions/new.php]: What's the rationale for CSS here?

File: interface/forms/care_plan/new.php [hunk 8]
```diff
@@ -148,99 +166,111 @@ function set_related(codetype, code, selector, codedesc) {
...
+                            <div class="clearfix"></div>
+                        </div>
+                    </div>
+
+                    <?php }
+                    ?>
+
+                </fieldset>
+                <div class="form-group">
+                    <div class="col-sm-12 text-center">
```
Comment (robertdown) [inline interface/forms/care_plan/new.php]: All of these Save/Cancel buttons will need to be right justified

File: interface/forms/care_plan/new.php [hunk 9]
```diff
@@ -148,99 +166,111 @@ function set_related(codetype, code, selector, codedesc) {
...
+                        <div class="btn-group" role="group">
+                            <a href="javascript:top.restoreSession();document.my_form.submit();" class="btn btn-default btn-save"><?php echo xlt('Save'); ?></a>
+                            <a href='<?php echo "$rootdir/patient_file/encounter/$returnurl";?>' class="btn btn-default btn-cancel" onclick="top.restoreSession()"><?php echo xlt('Don\'t Save'); ?></a>
```
Comment (robertdown) [inline interface/forms/care_plan/new.php]: Let's unify the language, a user can either "Save" a form or "Cancel". I don't like the "Don't Save" text. Also, remove `btn-default` and replace with `btn-link`

File: interface/forms/clinical_instructions/new.php [hunk 10]
```diff
@@ -36,28 +37,61 @@
...
+                        <fieldset>
+                        <legend class=""><?php echo xlt('Instructions'); ?></legend>
+                            <div class="form-group">
+                                <div class="col-sm-10 col-sm-offset-1">
+                                    <textarea name="instruction" id ="instruction"  class="form-control" cols="80" rows="5" ><?php echo text($check_res['instruction']); ?></textarea>
+                                </div>
+                            </div>
+                        </fieldset>
+                        <div class="form-group">
+                            <div class="col-sm-12 text-center">
```
Comment (robertdown) [inline interface/forms/clinical_instructions/new.php]: text-right

File: interface/forms/clinical_instructions/new.php [hunk 11]
```diff
@@ -36,28 +37,61 @@
...
+                                <div class="btn-group" role="group">
+                                    <a href="javascript:top.restoreSession();document.my_form.submit();" class="btn btn-default btn-save"><?php echo xlt('Save'); ?></a>
+                                    <a href='<?php echo "$rootdir/patient_file/encounter/$returnurl";?>' class="btn btn-default btn-cancel" onclick="top.restoreSession()"><?php echo xlt('Don\'t Save'); ?></a>
```
Comment (robertdown) [inline interface/forms/clinical_instructions/new.php]: Remove `btn-default` add `btn-link`, standardize text

File: interface/forms/dictation/new.php [hunk 12]
```diff
@@ -1,42 +1,77 @@
+<!-- Form generated from formsWiz -->
 <?php
-/**
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: This all needs to stay

File: interface/forms/dictation/new.php [hunk 13]
```diff
@@ -1,42 +1,77 @@
...
 include_once("$srcdir/api.inc");
 formHeader("Form: dictation");
 $returnurl = 'encounter_top.php';
 ?>
-<html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<html>
+<head>
+<?php Header::setupHeader('bootstrap');?>
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: No need for bootstrap here, it's autoloaded

File: interface/forms/dictation/new.php [hunk 14]
```diff
@@ -1,42 +1,77 @@
...
+
+<style type="text/css" title="mystyles" media="all">
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: Why styles?

File: interface/forms/dictation/new.php [hunk 15]
```diff
@@ -1,42 +1,77 @@
...
+                    <fieldset>
+                            <legend class=""><?php echo xlt('Additional Notes'); ?></legend>
+                                <div class="form-group">
+                                    <div class="col-sm-10 col-sm-offset-1">
+                                        <textarea name="additional_notes"   class="form-control" cols="80" rows="5" ></textarea>
+                                    </div>
+                                </div>
+                    </fieldset>
+                    <div class="form-group">
+                        <div class="col-sm-12 text-center">
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: text-right

File: interface/forms/dictation/new.php [hunk 16]
```diff
@@ -1,42 +1,77 @@
...
+                            <div class="btn-group" role="group">
+                                <a href="javascript:top.restoreSession();document.my_form.submit();" class="btn btn-default btn-save"><?php echo xlt('Save'); ?></a>
+                                <a href="<?php echo "$rootdir/patient_file/encounter/$returnurl";?>" class="btn btn-default btn-cancel" onclick="top.restoreSession()"><?php echo xlt('Don\'t Save'); ?></a>
```
Comment (robertdown) [inline interface/forms/dictation/new.php]: Remove `btn-default` add `btn-link`, standardize text

File: interface/forms/dictation/view.php [hunk 17]
```diff
@@ -1,44 +1,85 @@
+<!-- Forms generated from formsWiz -->
 <?php
-/**
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: This needs to stay

File: interface/forms/dictation/view.php [hunk 18]
```diff
@@ -1,44 +1,85 @@
+<!-- Forms generated from formsWiz -->
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: Where did this come from?

File: interface/forms/dictation/view.php [hunk 19]
```diff
@@ -1,44 +1,85 @@
...
+
+use OpenEMR\Core\Header;
 
 include_once("../../globals.php");
 $returnurl = 'encounter_top.php';
 ?>
 <html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<?php Header::setupHeader('bootstrap');?>
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: Pull bootstrap

File: interface/forms/dictation/view.php [hunk 20]
```diff
@@ -1,44 +1,85 @@
...
+
+<style type="text/css" title="mystyles" media="all">
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: Why styles?

File: interface/forms/dictation/view.php [hunk 21]
```diff
@@ -1,44 +1,85 @@
...
+                        <legend class=""><?php echo xlt('Additional Notes'); ?></legend>
+                            <div class="form-group">
+                                <div class="col-sm-10 col-sm-offset-1">
+                                    <textarea name="additional_notes"   class="form-control" cols="80" rows="5" ><?php echo text($obj{"additional_notes"});?></textarea>
+                                </div>
+                            </div>
+                </fieldset>
+                
+                    <div class="form-group">
+                        <div class="col-sm-12 text-center">
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: text-right

File: interface/forms/dictation/view.php [hunk 22]
```diff
@@ -1,44 +1,85 @@
...
+                            <div class="btn-group" role="group">
+                                <a href="javascript:top.restoreSession();document.my_form.submit();" class="btn btn-default btn-save"><?php echo xlt('Save'); ?></a>
+                                <a href="<?php echo "$rootdir/patient_file/encounter/$returnurl";?>" class="btn btn-default btn-cancel" onclick="top.restoreSession()"><?php echo xlt('Don\'t Save'); ?></a>
```
Comment (robertdown) [inline interface/forms/dictation/view.php]: Remove `btn-default` add `btn-link`, standardize text

File: interface/forms/procedure_order/new.php [hunk 23]
```diff
@@ -396,130 +396,179 @@ function validate(f) {
 });
```
Comment (robertdown) [inline interface/forms/procedure_order/new.php]: This page was already updated with bootstrap and modernized. Suggest keeping it as-is

File: interface/forms/reviewofs/new.php [hunk 24]
```diff
@@ -15,164 +15,827 @@
...
+
 include_once("../../globals.php");
 include_once("$srcdir/api.inc");
 formHeader("Form: reviewofs");
 $returnurl = 'encounter_top.php';
 ?>
 <html><head>
-<?php html_header_show();?>
-<link rel="stylesheet" href="<?php echo $css_header;?>" type="text/css">
+<?php Header::setupHeader('bootstrap');?>
```
Comment (robertdown) [inline interface/forms/reviewofs/new.php]: Don't need the bootstrap here

File: interface/forms/reviewofs/new.php [hunk 25]
```diff
@@ -15,164 +15,827 @@
...
+
+
+<style type="text/css" title="mystyles" media="all">
```
Comment (robertdown) [inline interface/forms/reviewofs/new.php]: Why styles?

File: interface/forms/reviewofs/new.php [hunk 26]
```diff
@@ -15,164 +15,827 @@
...
+                <fieldset>
+                        <legend class=""><?php echo xlt('Additional Notes');?></legend>
+                            <div class="form-group">
+                                <div class="col-sm-10 col-sm-offset-1">
+                                    <textarea name="additional_notes"   class="form-control" cols="80" rows="5" ></textarea>
+                                </div>
+                            </div>
+                </fieldset>
+                <div class="form-group">
+                    <div class="col-sm-12 text-center">
```
Comment (robertdown) [inline interface/forms/reviewofs/new.php]: text-right

File: interface/forms/reviewofs/new.php [hunk 27]
```diff
@@ -15,164 +15,827 @@
...
+                        <div class="btn-group" role="group">
+                            <a href="javascript:top.restoreSession();document.my_form.submit();"  class="btn btn-default btn-save"><?php echo xlt('Save')?></a>
+                            <a href="<?php echo "$rootdir/patient_file/encounter/$returnurl";?>" class="btn btn-default btn-cancel"  onclick="top.restoreSession()"><?php echo xlt('Don\'t Save');?></a>
```
Comment (robertdown) [inline interface/forms/reviewofs/new.php]: Remove `btn-default` add `btn-link`, standardize text

File: interface/forms/soap/templates/general_new.html (hunk 1 above)
Comment (robertdown) [inline interface/forms/soap/templates/general_new.html]: I think adding this to all the forms is very burdensome as it dictates specific styling to the form regardless of the theme selected. My recommendation is to remove all of these.

File: interface/forms/clinical_instructions/new.php (hunk 6 above)
Comment (zbig01) [inline interface/forms/clinical_instructions/new.php]: Yes, I realize. Will remove

File: interface/forms/clinical_instructions/new.php (hunk 10 above)
Comment (zbig01) [inline interface/forms/clinical_instructions/new.php]: The reason for centering is because I was using <div class="container"> which centers the whole page and visually the centered buttons looked better balanced than having it pulled to the right

File: interface/forms/soap/templates/general_new.html (hunk 1 above)
Comment (zbig01) [inline interface/forms/soap/templates/general_new.html]: I agree that internal stylesheets take away the main advantages of the external stylesheets. Removing this stylesheet at this juncture will revert styling of these pages to the default bootstrap style.  The most noticeable effect would be that it would break the style of display of blocks of elements in the form as the base bootstrap does not assign separate  background colors to legends and fieldsets. 
It is vital for the fieldsets and legends to be distinguishable from the background as all the forms that have been so styled break up the form into logical sections using fieldsets and legends that, in my opinion, makes for an easier user experience.
To see what I am talking about use light.css, inspect the page in chrome/firefox and delete this internal stylesheet.  Of course this will not affect flat red as I have included these styles in that external stylesheet ;)
 I have carefully styled the various elements in flat red to minimize the difference in display of various elements in pages styled using bootstrap and those without. Once everything gets converted to bootstrap the need for such overrides it will diminish.
These internal styles will likely end up in an external stylesheet that can be imported into the other external stylesheets  à la core.css. 
One thing you would have noticed in trying to make the forms responsive we need to use media queries. These  need to be individualized for the various forms and would likely require internal stylesheets.

File: interface/forms/care_plan/new.php (hunk 9 above)
Comment (zbig01) [inline interface/forms/care_plan/new.php]: I agree with unifying the language. Please let the  btn-dafault live for now.

**Comment by zbig01** [comment]
I will remove bootstrap form 
`<?php Header::setupHeader('bootstrap');?>.`
I will change all Don't save to Cancel. Will Insert DocBlock.  

I would like to leave in the internal stylesheet for now. 

Regarding btn-default vs btn-link please review the 12 pages styled in Encounter > Administration and Encounter > Clinical and the 5 additional pages styled in Fees before we axe btn-default

File: interface/forms/care_plan/new.php (hunk 9 above)
Comment (robertdown) [inline interface/forms/care_plan/new.php]: Can't let the `btn-default` stay as we've already got several other places in the codebase using the `btn-link` - the reason for this is separate classes offer the ability to de-emphasize the negative action button.

File: interface/forms/clinical_instructions/new.php (hunk 10 above)
Comment (robertdown) [inline interface/forms/clinical_instructions/new.php]: We're working to unify all the action buttons to be in the same location everywhere, we landed on the right side; this promotes continuity of the user experience. Lots of things can be manipulated by themes, but button placement has to remain in the same place.

File: interface/forms/care_plan/new.php (hunk 9 above)
Comment (bradymiller) [inline interface/forms/care_plan/new.php]: @robertdown ,
The btn-link does not seem to look very good on zbig01's theme. Rec testing and seeing what I mean on the other forms where btn-link is used. I think this may be because zbig01's buttons are colored(ie. more obvious than the buttons in light theme). Is there a way for a theme to override this btn-link styling for the de-emphasized buttons?

**Comment by robertdown** [comment]
We separate the positive and negative action classes to allow the theme to decide how to present the buttons. For instance, in the default theme, Light, we intentionally apply less visual weight to the negative action buttons because I want the user to be drawn to the positive action as that will be what they are searching for most of the time. If you wanted both classes to look the same just change https://github.com/openemr/openemr/blob/master/interface/themes/style_flat_red.css#L1895 from `.btn-default` to `.btn-default, .btn-link`. We can't have positive and negative actions to have the same class

**Comment by zbig01** [comment]
Good point will incorporate suggestion into flat red stylesheet

File: interface/forms/soap/templates/general_new.html (hunk 1 above)
Comment (bradymiller) [inline interface/forms/soap/templates/general_new.html]: @zbig01 and @robertdown ,
1. If we migrate legend and fieldset styling to the external stylesheets(this would be easy to do and would then allow per style sheet customizations of these elements), then would we be able to remove the internal stylesheets?
2. And regarding the media queries in styling, why is this needed. Doesn't bootstrap manage that by itself?

**Comment by zbig01** [comment]
As you saw with ROS the bootstrap media query only kicks in later at a smaller screen size. As each form is different, to keep them all responsive and still easy to use when 3 or 4 tabs are locked we will need slightly different media queries. Once again after more pages are styled a common media query may emerge that could be incorporated into what we can call the oemr-bs3-override.css  that can be imported into any stylesheet

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

File: interface/forms/soap/templates/general_new.html [hunk 28]
```diff
@@ -4,49 +4,79 @@
 <title>{xl t='SOAP'|escape:'html'}</title>
 
 {headerTemplate}
+{literal}
+ <style type="text/css" title="mystyles" media="all">
```
Comment (bradymiller) [inline interface/forms/soap/templates/general_new.html]: hi @zbig01 ,
When media is 'all', is that line even needed?

Comment (bradymiller) [inline interface/forms/soap/templates/general_new.html]: disregard above, since this is just part of the style tag

**Comment by bradymiller** [comment]
@robertdown and @zbig01 ,

Posted a proposal that solves issues with some compromises on both sides :)

Here's the commit for the SOAP form that I just posted to this PR:
https://github.com/openemr/openemr/pull/946/commits/d8af1f83c69bcb1144a39b4e476a029538643e1b

This will do a couple things.
1. Removes in page styling most of the time (except for special cases of course like when need to have fine control of media)
2. Supports both Cancel buttons (link or a button)
3. Pull buttons to the left which I think looks nice. To the right looks funny(because of the legend/fieldset spacing likely) and just seems so far away for these forms. And placing them in the middle just seems awkward which I think is because user interfaces don't usually do that.

If we can agree on this, then can use this strategy for the rest of the forms and bring OpenEMR into the golden bliss of responsive

-brady

**Comment by zbig01** [comment]
Brady 
Had a quick look at your code. In fact I have been working on some changes over the past few days and have incorporated several changes that address @robertdown concerns. 
I will be shortly be updating the PR.
Meanwhile look at these two articles, they might be of interet
[](https://www.smashingmagazine.com/2016/11/a-quick-guide-for-designing-better-buttons/)
[](https://uxplanet.org/designing-more-efficient-forms-structure-inputs-labels-and-actions-e3a47007114f)

Sorry for t

**Comment by robertdown** [comment]
As a reference, the tl;dr; version of those articles:

* Visual weights in action buttons are key
* Buttons should be placed inline with the input items (Which I am ok with)
* Form is processed faster when labels are on top of form control (Ok with this too, for the most part. The speed difference was comparable and sometimes screen space dictates left sided (Thought right-aligned) labels)
* Action and secondary buttons need to be separate.

Thanks for the links!

**Comment by zbig01** [comment]
Made changes to the various forms in Encounter > Clinical and Encounter> Administrative and committed to PR - Flat bootstrap rev 4 #938.

**Comment by bradymiller** [comment]
hi @zbig01 ,
Please work from the branch on this PR to focus on your solution for the included forms (and feel free to `git revert` my commit if it makes it easier). I don't have the resources to merge(these merges take a significant amount of resources) the same code into the current codebase multiple times :)
Once we have a agreed upon solution, then will work on bringing in the other form sets into the current codebase.
thanks,
-brady

**Comment by bradymiller** [comment]
Project has been continued on the following PR:
https://github.com/openemr/openemr/pull/993
